In [1]:
import pandas as pd
import numpy as np
import random
import matplotlib.pyplot as plt


# Test and Training Data Sets

In [2]:
#Establish Training Set
ar_train_x = np.array([[0.90, 0.87], 
              [1.31, 0.75], 
              [2.48, 1.14], 
              [0.41, 1.87], 
              [2.45, 0.52], 
              [2.54, 2.97], 
              [0.07, 0.09],
              [1.32, 1.96],
              [0.94, 0.34],
              [1.75, 2.21],
             ])
#Training Set Target values
ar_train_y = np.array([1.0, 1.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 1.0, 0.0])

#Establish Test Set
ar_test_x = np.array([[1.81, 1.02], 
              [2.36, 1.60], 
              [2.17, 2.08], 
              [2.85, 2.91], 
              [1.05, 1.93], 
              [2.32, 1.73], 
              [1.86, 1.31],
              [1.45, 2.19],
              [0.28, 0.71],
              [2.49, 1.52],
             ])

#Test Set Target Values
ar_test_y = np.array([0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0])

# Perceptron Class

In [3]:
#Create perceptron class
class Perceptron:
    
    #initialize perceptron class
    def __init__(self, weights, bias):
        self.weights = weights
        self.delta_weights = np.zeros_like(weights)
        self.bias = bias
        self.delta_bias = 0
        self.activity = 0
        self.activation = 0
        self.delta = 0
    
    #Calculate the activity value
    def calc_activity(self, inputs):
        self.activity = np.dot(inputs, self.weights) + self.bias
        #print("activity: ", self.activity)
    
    #Calculate the activation function given the activity value using sigmoid function
    def calc_activation(self, activity):
        self.activation = 1/(1+np.exp(-1*self.activity))
        #print("activation: ", self.activation)
    
    #Find the change needed in weights and bias
    def set_delta_weights(self, inputs, eta, target):
        self.delta = self.activation*(1-self.activation)*(target - self.activation)
        for i in range(len(inputs)):
            self.delta_weights[i] = inputs[i] * self.delta * eta
        self.delta_bias =  self.delta * eta
        #print("Bias: ", self.bias, " Delta: ", self.delta, " Eta: ", eta, " Delta Bias: ", self.delta_bias)
    
    #update weights and bias
    def update_weights(self):
        #print("Weights: ", self.weights)
        #print(self.delta_weights)
        #print("Bias: ", self.bias)
        #print(self.delta_bias)
        self.weights += self.delta_weights
        self.bias += self.delta_bias
        #print("Updated bias: ", self.bias)
    
    #Call necessary functions to update perceptron using perceptron delta function process
    def perceptron_delta(self, target, inputs):
        self.target = target
        self.calc_activity(inputs)
        self.calc_activation(self.activity)
        self.set_delta_weights(inputs, eta, target)
        self.update_weights()

# Initialize Single Layer Perceptron and Find Best Weights/Eta

In [4]:
#Define weights, eta values, iterations, and bias
weight_vector = [[-1.0, -1.0], [-0.5, -0.5], [0.5, 0.5], [1.0, 1.0]]
eta_values = [[0.1], [0.5], [1.0]]
iterations = 30
bias = 0

#Variables to capture best model
best_weight = 0
best_eta = 0
lowest_E = float('inf')

#for each set of weights
for g in range(0, len(weight_vector)):
    
    #for each eta value
    for h in range(0, len(eta_values)):
        #set eta
        eta = eta_values[h][0]
        
        #Create a Perceptron object
        Percep = Perceptron(weight_vector[g], bias)
        
        #train the network for specified number of iterations
        for i in range(0,iterations):
            total_error = 0
            
            #for every input pair
            for j in range (0, len(ar_train_x)):
                #Update the perceptron weights and bias based on the training output value
                Percep.perceptron_delta(ar_train_y[j], ar_train_x[j])
                
                #if final iteration, collect  total error
                #if i== iterations-1:
                    #total_error = total_error + Percep.delta
                total_error = total_error + (0.5)*(ar_train_y[j]-Percep.activation)**2
            #print("Average Error for iteration:", i+1," is ", round(abs(total_error/len(ar_train_x)),3))
            
            #print("Starting Weights: ", weight_vector[g], " Eta: ", eta, " Total Error: ", total_error )
            if(abs(total_error) < lowest_E):
                lowest_E=total_error
                best_weights = weight_vector[g]

                best_eta = eta
            #if i==29:
                #print("Start Weights", weight_vector[g], "Final Weights: ",Percep.weights, "Bias: ", Percep.bias, " Avg Error: ", total_error/10)

print("Best Weight: ", best_weights, " Best eta: ", best_eta, "Lowest E: ", lowest_E, ", when averaged is", lowest_E/10)

Best Weight:  [-1.0, -1.0]  Best eta:  1.0 Lowest E:  0.8304399711430172 , when averaged is 0.08304399711430172


# Initialize Single-Layer Perceptron with Best Weights/Eta

In [5]:
#Set weights and eta to best model
eta = best_eta
weight_vector = best_weights

## Retrain perceptron using Method 1 (30 iters, sequential I/O pairs)

In [6]:
#train perceptron with best model weights and eta
Percep = Perceptron(best_weights, bias)
for i in range(0,iterations):            
        #for every input pair
        for j in range (0, len(ar_train_x)):
            #Update the perceptron weights and bias based on the training output value
            Percep.perceptron_delta(ar_train_y[j], ar_train_x[j])

## Using trained model, make predictions on training data

In [7]:
#With trained model, use training value predictions to find optimal threshold:
predictions = []
for j in range (0, len(ar_train_x)): 
    Percep.calc_activation(Percep.calc_activity(ar_train_x[j]))
    pred = Percep.activation
    predictions.append(pred)
    print("Predicted: ", round(pred,3), " Actual:", ar_train_y[j]) 

Predicted:  0.582  Actual: 1.0
Predicted:  0.502  Actual: 1.0
Predicted:  0.072  Actual: 0.0
Predicted:  0.231  Actual: 0.0
Predicted:  0.246  Actual: 0.0
Predicted:  0.001  Actual: 1.0
Predicted:  0.964  Actual: 1.0
Predicted:  0.062  Actual: 0.0
Predicted:  0.812  Actual: 1.0
Predicted:  0.02  Actual: 0.0


## Find optimal threshold.

In [8]:
threshold = 0.0
best_accuracy = 0
best_threshold = 0.0

while threshold < 1.0:
    mapped_predictions = [0 if val < threshold else 1 for val in predictions]
    
    
    # First, calculate the number of correct predictions
    correct_predictions = sum([1 for a, b in zip(ar_train_y, mapped_predictions) if a == b])

    # Then, calculate the accuracy as a percentage of correct predictions
    accuracy = correct_predictions / len(ar_train_y) * 100
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_threshold = threshold
        
    threshold = threshold + 0.02

print("Best Accuracy:", best_accuracy, " when threshold is: ", round(best_threshold,4))

Best Accuracy: 90.0  when threshold is:  0.26


## Apply threshold and evaluate against Test set

In [9]:
predictions = []

#With trained model, use test values to evaluate
for j in range (0, len(ar_test_x)): 
    Percep.calc_activation(Percep.calc_activity(ar_test_x[j]))
    pred = Percep.activation
    predictions.append(pred)
    #print("Predicted: ", round(pred,3), " Actual:", ar_test_y[j]) 
    
mapped_predictions = [0 if val < best_threshold else 1 for val in predictions]
correct_predictions = sum([1 for a, b in zip(ar_test_y, mapped_predictions) if a == b])
accuracy = correct_predictions / len(ar_test_y) * 100
print(accuracy)

80.0


## A threshold of 0.26 would correctly match 80% of Test and 90% Train.

# Multi-Layer Case

## Define layer class that will be used for a 2 layer network

In [10]:
class Layer:
    #initialize layer class
    def __init__(self, layer_weight, eta = 0,  bias = None):
        self.layer_weight = layer_weight
        self.layer_bias = bias
        self.eta = eta
    
    def sigmoid(self, x):
        return 1/(1+np.exp(-x))
    
    def get_layer_output_vector(self, inputs):
        self.activity = np.dot(inputs ,self.layer_weight)+self.layer_bias
        self.output_vector = 1/(1+np.exp(-self.activity))
        return self.output_vector
    
    def get_error_vector(self, desired_output):
        self.littleE_vector = desired_output - self.output_vector
        return self.littleE_vector
    
    def calc_delta_k(self, target):
        #print("Target is ", target)
        #print("Output vector is: ", self.output_vector)
        self.littleE = target - self.output_vector
        #print("Little E is ", self.littleE)
        self.bigE = 0.5*self.littleE**2
        self.delta_k = self.littleE*(1-self.output_vector)*self.output_vector
        return self.delta_k
    
    def calc_delta_j(self, delta_k, next_layer_weights):
        #print(delta_k)
        #print(next_layer_weights)
        self.delta_j = (1-self.output_vector)*self.output_vector*(delta_k*next_layer_weights).T
        return self.delta_j
    
    def calc_delta_weights (self, delta, prev_layer_input):
        #print("---DELTA WEIGHT CALC----")
        #print("Self Weights", self.layer_weight)
        #print("shape ", self.layer_weight.shape)
        #print("eta", self.eta)
        #print("delta", delta)
        #print("Prev Layer input ", prev_layer_input)
        self.delta_weights = self.layer_weight + self.eta*delta*prev_layer_input
        self.delta_bias = self.layer_bias + self.eta*delta
        return self.delta_weights, self.delta_bias
    
    def update_layer_weights(self):
        self.layer_weight = self.delta_weights
        self.layer_bias = self.delta_bias
        

# Initialize Multi-Layer Network and Find Best Weights/Eta

In [11]:
#Define weights, eta values, iterations, and bias
weight_layer_1 = np.array([[[-1.0, -1.0], [-1.0, -1.0]], [[-0.5, -0.5], [-0.5, -.05]], [[0.5, 0.5], [0.5, 0.5]], [[1.0, 1.0], [1.0, 1.0]]])
weight_layer_2 = np.array([[-1.0, -1.0], [-0.5, -0.5], [0.5, 0.5], [1.0, 1.0]])

eta_values = [[random.uniform(0,1)], [random.uniform(0,1)], [random.uniform(0,1)]]
iterations = 30
bias = 0

#Variables to capture best model
best_weights1 = 0
best_weights2 = 0
best_eta = 0
lowest_E = float('inf')

#for each set of weights in layer 1
for f in range(0, len(weight_layer_1)):
    #Layer1=Layer(weight_layer_1[f].reshape(2,2), eta, bias)
    #print("Layer 1 weights: ", Layer1.layer_weight)
    
    #for each set of weights in layer 2
    for g in range (0, len(weight_layer_2)):
        
        #for each eta value in our list
        for h in range(0, len(eta_values)):
            #set eta, layer weights
            eta = eta_values[h][0]
            Layer1=Layer(weight_layer_1[f].reshape(2,2), eta, bias)
            #print("Layer 1 weights: ", Layer1.layer_weight)
            Layer2=Layer(weight_layer_2[g].reshape(2,1), eta, bias)
            #print("Layer 1 weights: ", Layer1.layer_weight)
            
            #train the network for specified number of iterations
            for i in range(0,iterations):
                total_error = 0
                #Train network for method 1:
                for j in range(0,len(ar_train_x)):
                    #Input1
                    #print("For input pair: ", j, " ,", ar_train_x[j])
                    Layer1.get_layer_output_vector(ar_train_x[j])
                    #print("Layer1  layer weight is: ", Layer1.layer_weight)
                    #print("Layer1 layer bias is: ", Layer1.layer_bias)
                    #print("Layer 1 activity vector is: ", Layer1.activity)
                    #print("Layer 1 output vector is: ", Layer1.output_vector)
                    Layer2.get_layer_output_vector(Layer1.output_vector)
                    #print("Layer2  layer weight is: ", Layer2.layer_weight)
                    #print("Layer2 layer bias is: ", Layer2.layer_bias)
                    #print("Layer2 activity vector is: ", Layer2.activity)
                    #print("Layer2 output vector is: ", Layer2.output_vector)
                    #print("Input 1, Iteration: ", i, " Layer output ", Layer2.output_vector)
                    Layer2.calc_delta_k(ar_train_y[j])
                    #print("Layer2 little E vector is: ", Layer2.littleE)
                    #print("Layer2 Big E vector is: ", Layer2.bigE)
                    #print("Layer2 delta k vector is: ", Layer2.delta_k)
                    Layer2.calc_delta_weights(Layer2.delta_k, Layer1.output_vector.reshape(2,1))
                    #print("Layer 2 delta weights :", Layer2.delta_weights)
                    #print("layer 2 delta bias: ", Layer2.delta_bias)
                    Layer1.calc_delta_j(Layer2.delta_k, Layer2.layer_weight)
                    #print("Layer1 delta j: ", Layer1.delta_j)
                    Layer1.calc_delta_weights(Layer1.delta_j, ar_train_x[j])
                    #print("Layer 1 delta weights :", Layer1.delta_weights)
                    #print("layer 1 delta bias: ", Layer1.delta_bias)
                    Layer1.update_layer_weights()
                    #print("Layer 1 updated weights :", Layer1.layer_weight)
                    #print("layer 1 updated bias: ", Layer1.layer_bias)
                    Layer2.update_layer_weights()
                    #print("Layer 2 updated weights :", Layer2.layer_weight)
                    #print("layer 2 updated bias: ", Layer2.layer_bias)
                    total_error = total_error + Layer2.bigE
                    #print("Total error for iteration ", i, " and input pair ", j, " is ", Layer2.bigE)
                    #print("------Break-------")

                if(abs(total_error) < lowest_E):
                    lowest_E=total_error
                    best_weights1 = weight_layer_1[f]
                    best_weights2 = weight_layer_2[g]
                    best_eta = eta
                
print("Best Weight for layer 1: ", best_weights1, "Best weights for layer 2:", best_weights2,  " Best eta: ", best_eta, "Lowest E: ", lowest_E, ", when averaged is ",lowest_E/10)

Best Weight for layer 1:  [[-1. -1.]
 [-1. -1.]] Best weights for layer 2: [1. 1.]  Best eta:  0.6555333371303738 Lowest E:  [[0.9438601]] , when averaged is  [[0.09438601]]


# Initialize Multi-Layer Network with Best Weights/Eta

In [12]:
#Set weights and eta to best model
eta = best_eta
weight_vector_1 = best_weights1
weight_vector_2 = best_weights2
bias = 0

Layer1=Layer(weight_vector_1.reshape(2,2), eta, bias)
Layer2=Layer(weight_vector_2.reshape(2,1), eta, bias)

## Retrain Network using Method 1 (30 iters, sequential I/O pairs)

In [13]:
#train the network with the best eta and weight values
for i in range(0,iterations):
    #Train network for method 1:
    for j in range(0,len(ar_train_x)):
        Layer1.get_layer_output_vector(ar_train_x[j])
        Layer2.get_layer_output_vector(Layer1.output_vector)
        Layer2.calc_delta_k(ar_train_y[j])
        Layer2.calc_delta_weights(Layer2.delta_k, Layer1.output_vector.reshape(2,1))
        Layer1.calc_delta_j(Layer2.delta_k, Layer2.layer_weight)
        Layer1.calc_delta_weights(Layer1.delta_j, ar_train_x[j])
        Layer1.update_layer_weights()
        Layer2.update_layer_weights()

## Using trained model, make predictions on training data

In [14]:
#With trained model, use training value predictions to find optimal threshold:
predictions = []
for j in range (0, len(ar_train_x)): 
    Layer1.get_layer_output_vector(ar_train_x[j])
    Layer2.get_layer_output_vector(Layer1.output_vector)
    pred = Layer2.output_vector
    predictions.append(pred)
    print("Predicted: ", np.round(pred,3), " Actual:", ar_train_y[j]) 

Predicted:  [[0.556]]  Actual: 1.0
Predicted:  [[0.502]]  Actual: 1.0
Predicted:  [[0.353]]  Actual: 0.0
Predicted:  [[0.467]]  Actual: 0.0
Predicted:  [[0.392]]  Actual: 0.0
Predicted:  [[0.317]]  Actual: 1.0
Predicted:  [[0.868]]  Actual: 1.0
Predicted:  [[0.37]]  Actual: 0.0
Predicted:  [[0.664]]  Actual: 1.0
Predicted:  [[0.341]]  Actual: 0.0


## Find optimal threshold.

In [15]:
threshold = 0.0
best_accuracy = 0
best_threshold = 0.0

while threshold < 1.0:
    mapped_predictions = [0 if val < threshold else 1 for val in predictions]
    
    
    # First, calculate the number of correct predictions
    correct_predictions = sum([1 for a, b in zip(ar_train_y, mapped_predictions) if a == b])

    # Then, calculate the accuracy as a percentage of correct predictions
    accuracy = correct_predictions / len(ar_train_y) * 100
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_threshold = threshold
        
    threshold = threshold + 0.02

print("Best Accuracy:", best_accuracy, " when threshold is: ", round(best_threshold,4))

Best Accuracy: 90.0  when threshold is:  0.48


## Apply threshold and evaluate against Test set

In [16]:
predictions = []

#With trained model, use test values to evaluate
for j in range (0, len(ar_test_x)): 
    Layer1.get_layer_output_vector(ar_test_x[j])
    Layer2.get_layer_output_vector(Layer1.output_vector)
    pred = Layer2.output_vector
    predictions.append(pred)
    
mapped_predictions = [0 if val < best_threshold else 1 for val in predictions]
correct_predictions = sum([1 for a, b in zip(ar_test_y, mapped_predictions) if a == b])
accuracy = correct_predictions / len(ar_test_y) * 100
print(accuracy)

80.0


## A threshold of 0.48 would correctly match 80% of Test and 90% Train.

# BREAK  - Still working through what we would need to visualize.  ROCs, Specificity vs Sensitivity, etc.